### PR TITLE
[core][state] Task events backend for state API and profiling (events emitting changes) [1/2] 

### PR DESCRIPTION
**Pull request by rickyyx**
Signed-off-by: rickyyx <rickyx@anyscale.com>

## Why are these changes needed?

For details of the design and background see this [doc](https://docs.google.com/document/d/1l7GAbEoVbZszWOS2V1voSq5YiPzTgblK9dmtmnWb_tM/edit)

**TL;DR is we will emit task state changes from core workers/raylets to GCS in a similar way of how task profiling is currently done. This will allow persistence of task states to be queried by clients like State APIs or Dashboard.**

#### In this PR: 
Changes are mainly on writing task events to GCS to get performance on par. Read side changes will be in the next PR. 

1. Added a `TaskEventBuffer` class which serves as an abstraction to store task events, and push those events to GCS in batches periodically.  Each `CoreWorker` will own one single `TaskEventBuffer`, and events (both task status change events and profiling events) will be added to a local in-memory buffer. The data being pushed is defined in `TaskEventData`. 
   - The `TaskEventBuffer` also owns its own GCS client and io thread, which is independent from the main io_contexts. 
3. Added a GCS sergice `GcsTaskManager` that handles the reporting of such events from workers. 
   - The service has its own io thread and io context that's isolated from other GCS service as well. 
   - When a certain number of task events reached in the service, the service will start dropping older events approximately, where a task event is older if it was reported earlier than the others. 

When this PR is merged, with `RAY_task_event_report_interval_ms = 0` as default, the feature is turned off by default so there shouldn't be any regression. 

#### In the next PR: 
- Porting of profiling events query and State API task querying. 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/core_worker/core_worker.cc [hunk 1]
```diff
@@ -3237,13 +3273,6 @@ void CoreWorker::HandleGetCoreWorkerStats(rpc::GetCoreWorkerStatsRequest request
     task_manager_->AddTaskStatusInfo(stats);
   }
 
-  if (request.include_task_info()) {
```
Comment (rickyyx) [inline src/ray/core_worker/core_worker.cc:3240]: todo: this needs to be reverted.

File: src/ray/common/task/task_spec.h [hunk 2]
```diff
@@ -159,7 +159,7 @@ static inline rpc::ObjectReference GetReferenceForActorDummyObject(
 class TaskSpecification : public MessageWrapper<rpc::TaskSpec> {
  public:
   /// Construct an empty task specification. This should not be used directly.
-  TaskSpecification() {}
+  TaskSpecification() { ComputeResources(); }
```
Comment (rickyyx) [inline src/ray/common/task/task_spec.h:162]: Used only in test.

File: src/ray/core_worker/profiling.h [hunk 3]
```diff
@@ -20,6 +20,7 @@
 #include "ray/common/asio/instrumented_io_context.h"
```
Comment (rickyyx) [inline src/ray/core_worker/profiling.h:20]: Profiling frontend porting will be in the next PR.

File: src/ray/common/ray_config_def.h [hunk 4]
```diff
@@ -458,6 +458,17 @@ RAY_CONFIG(uint64_t, gcs_service_address_check_interval_milliseconds, 1000)
 /// The batch size for metrics export.
 RAY_CONFIG(int64_t, metrics_report_batch_size, 100)
 
+/// The interval duration for which task state events will be reported to GCS.
+/// The reported data should only be used for observability.
+/// Setting the value to 0 to disable the task event recording and reported.
+/// TODO(rickyx): Turn this off before merging
```
Comment (rickyyx) [inline src/ray/common/ray_config_def.h:464]: to be turned off - turning on now for runnign release tests and CI.

**Review by ericl — CHANGES_REQUESTED** [review]
Per our general policy to keep PRs <500 lines when possible, could we split this PR into a few chunks? Perhaps the utility data structures could be reviewed first, prior to the code plumbing.

**Comment by rickyyx** [comment]
> Per our general policy to keep PRs <500 lines when possible, could we split this PR into a few chunks? Perhaps the utility data structures could be reviewed first, prior to the code plumbing.

Ah Ok, let me do that instead. Thought I could try merging something working more end to end in. I think I could probably do the below: 
1. the RPC changes 
2. The `TaskEventBuffer` code and its unittests 
3. The GCS side changes. 

Each of the above should be small enough for an easier review I believe.

**Pull request closed without merging**